### PR TITLE
LibCore+Userland: Make IODevice::write return ErrorOr<void>

### PR DESCRIPTION
**Pull request by RasmusNylander**
This makes IODevice::write(StringView) and
IODevice::write(const u8*, int size) return ErrorOr<void> instead of a
boolean indicating success or failure.

Other than simply being inherently somewhat more pleasing, as the
boolean return merely functioned to approximate its new form, this also
forces callers to acknowledge the possibility of errors and actually
handle them.

Where easy and convenient, mostly where the caller already returned
ErrorOr, the IODevice::write calls have simply been wrapped in a TRY().
Generally, where the caller already were handling errors but did not
return ErrorOr, the check to see whether IODevice::write returned true
was replace with a check to see whether IODevice::write returned an
error. Elsewhere, they were for the most part just wrapped in a MUST().

A few places, using AK::StringBuilder, the caller was changed ever so
slightly to contain fewer calls to write.

This is my first semi-real commit and a rather snazzy one at that, if I
do say so myself. :^)

### CODE REVIEW

**Comment by RasmusNylander** [comment]
I knew I had forgotten something! Linting will have to wait until after my exam - possibly until tomorrow.

File: Userland/Games/Chess/ChessWidget.cpp [hunk 1]
```diff
@@ -634,46 +634,49 @@ bool ChessWidget::export_pgn(StringView export_path) const
...
     auto& file = *file_or_error.value();
 
+    StringBuilder builder;
     // Tag Pair Section
-    file.write("[Event \"Casual Game\"]\n");
-    file.write("[Site \"SerenityOS Chess\"]\n");
-    file.write(String::formatted("[Date \"{}\"]\n", Core::DateTime::now().to_string("%Y.%m.%d")));
-    file.write("[Round \"1\"]\n");
+    builder.append("[Event \"Casual Game\"]\n");
+    builder.append("[Event \"Casual Game\"]\n");
```
Comment (BenWiederhake) [inline Userland/Games/Chess/ChessWidget.cpp:640]: Possible typo? Before this commit, this line was output only once.

File: Userland/Games/Chess/ChessWidget.cpp [hunk 2]
```diff
@@ -634,46 +634,49 @@ bool ChessWidget::export_pgn(StringView export_path) const
...
+    builder.append(String::formatted("[Black \"{}\"]\n", m_side == Chess::Color::Black ? player1 : player2));
 
-    file.write(String::formatted("[Result \"{}\"]\n", Chess::Board::result_to_points(m_board.game_result(), m_board.turn())));
-    file.write("[WhiteElo \"?\"]\n");
-    file.write("[BlackElo \"?\"]\n");
-    file.write("[Variant \"Standard\"]\n");
-    file.write("[TimeControl \"-\"]\n");
-    file.write("[Annotator \"SerenityOS Chess\"]\n");
-    file.write("\n");
+    builder.append(String::formatted("[Result \"{}\"]\n", Chess::Board::result_to_points(m_board.game_result(), m_board.turn())));
```
Comment (BenWiederhake) [inline Userland/Games/Chess/ChessWidget.cpp:651]: For future reference: We also have `StringBuilder::appendff`, which saves an allocation. I don't think it matters here, but now you know :D

File: Userland/Libraries/LibAudio/WavWriter.cpp [hunk 3]
```diff
@@ -40,10 +40,11 @@ void WavWriter::set_file(StringView path)
     m_finalized = false;
 }
 
-void WavWriter::write_samples(const u8* samples, size_t size)
+ErrorOr<void> WavWriter::write_samples(const u8* samples, size_t size)
```
Comment (BenWiederhake) [inline Userland/Libraries/LibAudio/WavWriter.cpp:43]: This is perfectly fine.

But now that I see this function signature, I'd like to bait someone into making this accept a `Span<u8>`, just to prevent weird memory issues. (That would be a separate PR. You can immediately mark this "resolved".)

File: Userland/Libraries/LibCore/IODevice.cpp [hunk 4]
```diff
@@ -275,15 +275,19 @@ bool IODevice::truncate(off_t size)
     return true;
 }
 
-bool IODevice::write(const u8* data, int size)
+ErrorOr<void> IODevice::write(const u8* data, int size)
```
Comment (BenWiederhake) [inline Userland/Libraries/LibCore/IODevice.cpp:278]: This improves the situation. Great!

… Can I bait you into another PR after this one? It would be nice if we could report the amount of bytes actually written, so something like `ErrorOr<size_t>` would be nice. This way, doing non-blocking I/O could be made much more reliable. Note that this is going to be a huge endeavor, though.

File: Userland/Libraries/LibWebSocket/Impl/TCPWebSocketConnectionImpl.cpp [hunk 5]
```diff
@@ -44,7 +44,7 @@ void TCPWebSocketConnectionImpl::connect(ConnectionInfo const& connection)
 
 bool TCPWebSocketConnectionImpl::send(ReadonlyBytes data)
 {
-    return m_socket->write(data);
+    return m_socket->write(data).is_error();
```
Comment (BenWiederhake) [inline Userland/Libraries/LibWebSocket/Impl/TCPWebSocketConnectionImpl.cpp:47]: Do you intend to work on `TCPWebSocketConnectionImpl::send` next? If not, please add something like:
`// TODO: Propagate errno code`

**Review by BenWiederhake — COMMENTED** [review]
Lovely! One step closer to useful error reporting :D

Please add a `// TODO` comment around the place that I indicated, and consider using `StringBuilder::appendff` to avoid pointless allocations and copies. And of course I'm trying to bait you into writing more PRs after this one ;)
Overall this is very nice work :)

**Comment by creator1creeper1** [comment]
Good changes, although we are trying to move away from `IODevice`. See #9977. @sin-ack

File: Userland/Libraries/LibHTTP/HttpJob.cpp [hunk 6]
```diff
@@ -108,7 +108,7 @@ bool HttpJob::eof() const
 
 bool HttpJob::write(ReadonlyBytes bytes)
 {
-    return m_socket->write(bytes);
+    return m_socket->write(bytes).is_error();
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibHTTP/HttpJob.cpp:111]: I think this should be inverted?

File: Userland/Libraries/LibWebSocket/Impl/TCPWebSocketConnectionImpl.cpp (hunk 5 above)
Comment (AtkinsSJ) [reply Userland/Libraries/LibWebSocket/Impl/TCPWebSocketConnectionImpl.cpp:47]: Unrelated but same line: I think this should be inverted? Assuming that the return code means success.

File: Userland/Games/Chess/ChessWidget.cpp (hunk 1 above)
Comment (RasmusNylander) [reply Userland/Games/Chess/ChessWidget.cpp:640]: Certainly is!

File: Userland/Games/Chess/ChessWidget.cpp (hunk 2 above)
Comment (RasmusNylander) [reply Userland/Games/Chess/ChessWidget.cpp:651]: Oh that's pretty neat! I went ahead and implemented it anyway - it looks a little nicer :^)

File: Userland/Libraries/LibAudio/WavWriter.cpp (hunk 3 above)
Comment (RasmusNylander) [reply Userland/Libraries/LibAudio/WavWriter.cpp:43]: That sound interesting, I might look into that, now IODevice is deprecated. I make no promises though

File: Userland/Libraries/LibCore/IODevice.cpp (hunk 4 above)
Comment (RasmusNylander) [reply Userland/Libraries/LibCore/IODevice.cpp:278]: If it wasn't just depreacted, definitely! I supose this fancy new Stream-thing might benifit from something similar though. If it doesn't already have it, that is - I haven't looked at it yet.

File: Userland/Libraries/LibWebSocket/Impl/TCPWebSocketConnectionImpl.cpp (hunk 5 above)
Comment (RasmusNylander) [reply Userland/Libraries/LibWebSocket/Impl/TCPWebSocketConnectionImpl.cpp:47]: I did not; it has been added

File: Userland/Libraries/LibHTTP/HttpJob.cpp (hunk 6 above)
Comment (RasmusNylander) [reply Userland/Libraries/LibHTTP/HttpJob.cpp:111]: Definitely should!

File: Userland/Libraries/LibCpp/Tests/parser/strace.cpp [hunk 7]
```diff
@@ -135,7 +135,7 @@ int main(int argc, char** argv)
             arg3,
             res);
 
-        if (!trace_file->write(string)) {
+        if (trace_file->write(string).is_error()) {
```
Comment (davidot) [inline Userland/Libraries/LibCpp/Tests/parser/strace.cpp:138]: It seems this file is used in testing LibCpp and depends on the exact contents.
You probably need to regenerate the `Userland/Libraries/LibCpp/Tests/parser/strace.ast` file for the tests to pass.
I have quickly looked around but am not sure how to do that.
Perhaps cc @itamar8910 knows more about this

Comment (itamar8910) [reply Userland/Libraries/LibCpp/Tests/parser/strace.cpp:138]: Sure, try running `cpp-parser /usr/src/serenity/Libraries/LibCpp/Tests/parser/strace.cpp > strace.ast` and then replace `Userland/Libraries/LibCpp/Tests/parser/strace.ast` with the result.

File: Userland/Applications/KeyboardMapper/KeyboardMapperWidget.cpp [hunk 8]
```diff
@@ -203,8 +203,8 @@ void KeyboardMapperWidget::save_to_file(StringView filename)
         return;
     }
 
-    bool result = file->write(file_content);
-    if (!result) {
+    auto result = file->write(file_content);
+    if (result.is_error()) {
         int error_number = errno;
```
Comment (bgianfo) [inline Userland/Applications/KeyboardMapper/KeyboardMapperWidget.cpp:208]: Shouldn't be use the error from the result, and not assume `errno` is still the same error?

Comment (RasmusNylander) [reply Userland/Applications/KeyboardMapper/KeyboardMapperWidget.cpp:208]: It probably should, but i think that would kind of be outside the scope of this PR. Also, it was in changing things like this in KeyboardMapper that I got distracted and started working on IODevice, so I don't think it really makes a difference - KeyboardMapperWidget will soon enough no longer be responsible for handling error and will simply propagate them instead.

File: Userland/Libraries/LibCpp/Tests/parser/strace.cpp (hunk 7 above)
Comment (RasmusNylander) [reply Userland/Libraries/LibCpp/Tests/parser/strace.cpp:138]: So, I tried doing this and could only get it to work inside of Serenity itself. This would be fine if I had any clue as to how to export the resulting file - after a lot of different attempts, I ended up trying to send it through the internet - pastebin it, email it to myself, and so on - but the browser wasn't really up for the task. 

Honestly, at this point I find my enthusiasm for this project almost non-existant; I think it would be for the best if I were to simply abandon this PR. I was working on cleaning up the KeyboardMapper when I got distracted by this. I think that I'll continue work on that instead

Comment (itamar8910) [reply Userland/Libraries/LibCpp/Tests/parser/strace.cpp:138]: It's too bad that you had this experience. Next time please feel free to talk about technical problems you're having on Discord.

The method I use for transferring files out of serenity is `mount`ing the _disk_image on the host.

Comment (RasmusNylander) [reply Userland/Libraries/LibCpp/Tests/parser/strace.cpp:138]: Discord is a little intimidating though :^)

I did try to mount it but I couldn't get it to work. I have no idea why - I just tried again and this time it worked just fine. Imma just blame the fact that it's linux
